summingbird-core-test/src/test/scala/com/twitter/summingbird/planner/DagOptimizerTest.scala

```diff
@@ -0,0 +1,153 @@
+  implicit def sink1: Memory#Sink[Int] = ((_) => Unit)
+  implicit def sink2: Memory#Sink[(Int, Int)] = ((_) => Unit)
+
+  def genProducer: Gen[Producer[Memory, _]] = oneOf(genProd1, genProd2, summed)
+
+  test("DagOptimizer round trips") {
+    forAll { p: Producer[Memory, Int] =>
+      val dagOpt = new DagOptimizer[Memory] { }
+
+      assert(dagOpt.toLiteral(p).evaluate == p)
+    }
+  }
+
+  val dagOpt = new DagOptimizer[Memory] { }
+
+  test("ExpressionDag fanOut matches DependantGraph") {
+    forAll(genProducer) { p: Producer[Memory, _] =>
+      val expDag = dagOpt.expressionDag(p)._1
+
+      // the expression considers an also a fanout, so
+      // we can't use the standard Dependants, we need to
+      // us parentsOf as the edge function
+      val deps = new DependantGraph[Producer[Memory, Any]] {
+        override lazy val nodes: List[Producer[Memory, Any]] = Producer.entireGraphOf(p)
+        override def dependenciesOf(p: Producer[Memory, Any]) = Producer.parentsOf(p)
+      }
+
+      deps.nodes.foreach { n =>
+        deps.fanOut(n) match {
+          case Some(fo) => assert(expDag.fanOut(n) == fo)
+          case None => fail(s"node $n has no fanOut value")
+        }
+      }
+    }
+  }
+
+
+  val allRules = {
+    import dagOpt._
+
+    List(RemoveNames,
+      RemoveIdentityKeyed,
+      FlatMapFusion,
+      OptionMapFusion,
+      OptionToFlatMap,
+      KeyFlatMapToFlatMap,
+      FlatMapKeyFusion,
+      ValueFlatMapToFlatMap,
+      FlatMapValuesFusion,
+      FlatThenOptionFusion,
+      DiamondToFlatMap,
+      MergePullUp,
+      AlsoPullUp)
+  }
+
+  val genRule: Gen[Rule[dagOpt.Prod]] =
+    for {
+      n <- Gen.choose(1, allRules.size)
+      rs <- Gen.pick(n, allRules) // get n randomly selected
+    } yield rs.reduce(_.orElse(_))
+
+  test("Rules are idempotent") {
+    forAll(genProducer, genRule) { (p, r) =>
+      val once = dagOpt.optimize(p, r)
+      val twice = dagOpt.optimize(once, r)
+      assert(once == twice)
+    }
+  }
+
+  test("fanOut matches after optimization") {
+
+    forAll(genProducer, genRule) { (p, r) =>
+
+      val once = dagOpt.optimize(p, r)
+
+      val expDag = dagOpt.expressionDag(once)._1
+      // the expression considers an also a fanout, so
+      // we can't use the standard Dependants, we need to
+      // us parentsOf as the edge function
+      val deps = new DependantGraph[Producer[Memory, Any]] {
+        override lazy val nodes: List[Producer[Memory, Any]] = Producer.entireGraphOf(once)
+        override def dependenciesOf(p: Producer[Memory, Any]) = Producer.parentsOf(p)
+      }
+
+      deps.nodes.foreach { n =>
+        deps.fanOut(n) match {
+          case Some(fo) => assert(expDag.fanOut(n) == fo, s"node: $n, in optimized: $once")
+          case None => fail(s"node $n has no fanOut value")
+        }
+      }
+    }
+
+  }
+
+  test("test some idempotency specific past failures") {
```

Comment:
I don't know what to say. This was a hand minimized example (it took me about an hour) from a failure case found by scalacheck. Since the failures were quite rare, it took a long time to even find a failure with scalacheck, so once I found it, I wanted to test that failure every time. 

That's what I mean by "specific past failures".

I understand (somewhat) why this failed now, but I couldn't easily generate another that would also show the bug.

Can you suggest some specific text you would like to see me add?